src/persist-client/src/lib.rs

```diff
@@ -381,6 +381,42 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn fetch_upper() -> Result<(), Box<dyn std::error::Error>> {
+        mz_ore::test::init_logging();
+
+        let data = vec![
+            (("1".to_owned(), "one".to_owned()), 1, 1),
+            (("2".to_owned(), "two".to_owned()), 2, 1),
+        ];
+
+        let client = new_test_client().await?;
+
+        let shard_id = ShardId::new();
+
+        let (mut write1, _read1) = client
+            .open::<String, String, u64, i64>(NO_TIMEOUT, shard_id)
+            .await?;
+
+        let (mut write2, _read2) = client
+            .open::<String, String, u64, i64>(NO_TIMEOUT, shard_id)
+            .await?;
+
+        let res = write1.append_slice(&data[..], 3).await?;
+        assert_eq!(res, Ok(()));
+
+        let write2_upper = write2.upper().clone();
+        let global_upper = write2.fetch_upper(NO_TIMEOUT).await?;
+
+        // The writer upper should not advance if another writer advances.
```

Comment:
should the `write2.upper()` call be moved after the `fetch_upper` one? as-in, I'm not sure this test is covering what the comment says. IMO it's fine to just inline it in the assert_eq

Reply:
ah of course! for some reason I was imagining a situation where we were worried that fetching the state had some side effect of advancing the upper, but looking at it again with fresh eyes, I agree that it was correct before. oh well, now it covers both I guess :)